### PR TITLE
For #15: 0.4.0 Pre Release branch

### PR DESCRIPTION
**Pull request by wokket**
This is a body of work that we could consider complete, or have significant work still to do.

I'm opening this PR to give you an idea of where I'm headed with moving the awesome query support out of the string indexer, and possibly into it's own module somehow.

### CODE REVIEW

**Comment by wokket** [comment]
As it stands the two string indexers have been replaced with `query(&str)` and `query_by_string(String)` that just delegates to `query`, with the functionality of the functions unchanged.

**Comment by sempervictus** [comment]
Went over it, and suggest a compromise: can we leave the Index implementations, aliasing them to the functional one or visa-versa?
Without removing the Index impl, we simply need:
```rust
rust-hl7(feature/query_by)$ git diff
diff --git i/src/message.rs w/src/message.rs
index 7b82b21..337b47e 100644
--- i/src/message.rs
+++ w/src/message.rs
@@ -97,6 +97,14 @@ impl<'a> Message<'a> {
     pub fn as_str(&self) -> &'a str {
         self.source
     }
+
+    fn query(&self, idx: &str) -> &'a str {
+        self[idx]
+    }
+
+    fn query_by_string(&self, idx: String) -> &'a str {
+        self[idx]
+    }
 }
 
 impl<'a> Display for Message<'a> {
```

**Comment by wokket** [comment]
This PR has ended up with a bunch of stuff not related to #15 in it, so I'm repurposing it into the pre-0.4 PR 👍

**Comment by wokket** [comment]
@sempervictus  Anything in here you'd like tweaked? Otherwise I'll merge this to master and push out a 0.4 release 👍

**Comment by sempervictus** [comment]
I think we're good to merge for the most part. On the `TODO` bits - can we deref one level of the `&&str` to be `&str` somehow?
Otherwise i'm using roughly this state for prod work for a bit now and its _awesome_ compared to the options we had on the table prior (either slow and buggy or very slow and less buggy).

**Comment by wokket** [comment]
tldr: The `&&str` stuff is a result of the Indexer trait, with real reasons (that I don't entirely understand I have to admit) and have burnt an hour on already, so I'm leaving those as ToDo's for the moment, but that may just be the way it is.

Trying to re-add string based indexers now... are you specifically using the string indexers on `Message` only, or do you have code calling it on segments etc as well?

**Comment by sempervictus** [comment]
I have a bit of code that indexes into things based on what it received from a channel (Message, Segment, or Field) - so its handy to have at the lower tiers, but i can rewrite that to use the `usize` indexing approach.

**Comment by wokket** [comment]
Re-adding the string indexers was also a nightmare due to the Indexer trait typing, so I'm putting this on hold for 0.5.  Release 0.4 has just been published straight from master for you.
